packages/apify/src/actor.ts

```diff
@@ -609,13 +617,30 @@ export class Actor<Data extends Dictionary = Dictionary> {
      *
      * @param item Object or array of objects containing data to be stored in the default dataset.
      * The objects must be serializable to JSON and the JSON representation of each object must be smaller than 9MB.
+     * @param eventName If provided, the method will attempt to charge for the event for each pushed item.
      * @ignore
      */
-    async pushData(item: Data | Data[]): Promise<void> {
+    async pushData(item: Data | Data[]): Promise<void>;
+    async pushData(item: Data | Data[], eventName: string): Promise<ChargeResult>;
+    async pushData(item: Data | Data[], eventName?: string | undefined): Promise<ChargeResult | void> {
         this._ensureActorInit('pushData');
 
+        const chargeResult = eventName !== undefined
+            ? await this.charge({ eventName, count: Array.isArray(item) ? item.length : 1 })
+            : undefined;
+
         const dataset = await this.openDataset();
-        return dataset.pushData(item);
+
+        if (chargeResult?.eventChargeLimitReached) {
+            const items = Array.isArray(item) ? item : [item];
+            await dataset.pushData(items.slice(0, chargeResult.chargedCount));
```

Comment:
@halvko had a good point that in the super rare case when the Actor crashes in between the `.charge` and `.pushData` calls, it would be better to have pushData first and only after finish the charge. Considering we are talking few milliseconds, this will be super rare and it will complicate the code a bit but it would technically be more correct.

Reply:
My intuition would be that it's a bigger problem if we have items in dataset that were not in fact successfully charged than the opposite. Does anyone else want to chime in?

Reply:
My intuition is that customer is much more likely to be pissed off that they paid for no value, than a developer that would provide free data for a customer :)

Reply:
Ah, I lost context. I thought I was looking at the code responsible for the charge-audit dataset :facepalm:. You are correct, I'll try and fix this.

Reply:
FYI: By moving the push before the charge, we introduced a small race condition where as we are close to the max budget, more pushes can happen concurrently and we would push a few more items than the limit. I think this might happen maybe in 1/1000 runs, maybe even less (depends on the size of the data push which determines the latency of it).

I thought a bit about fixes for this and haven't come up with anything elegant. You could have a new `bookFutureCharge` method to pre-register events but it would introduce a new complex API. Or we could patch it on the `Actor` class bypassing the manager and just handle in the `pushData`. All of it introduces extra complexities. So for now I'm parking that thought because I don't think it is a big deal.